src/providers/wfs/qgswfsprovider.cpp

```diff
@@ -1285,43 +1268,26 @@ bool QgsWFSProvider::describeFeatureType( QString &geometryAttribute, QgsFields
   fields.clear();
 
   QgsWFSDescribeFeatureType describeFeatureType( mShared->mURI );
-  bool bUsePlural = false;
-  QByteArray response;
-  for ( int i = 0; i < 2; i++ )
-  {
-    if ( !describeFeatureType.requestFeatureType( mShared->mWFSVersion,
-         mShared->mURI.typeName(), bUsePlural ) )
-    {
-      QgsMessageLog::logMessage( tr( "DescribeFeatureType network request failed for url %1: %2" ).
-                                 arg( dataSourceUri(), describeFeatureType.errorMessage() ), tr( "WFS" ) );
-      return false;
-    }
 
-    response = describeFeatureType.response();
-    // "http://geoportal.samregion.ru/wfs12?SERVICE=WFS&REQUEST=DescribeFeatureType&TYPENAME=EC_1_132&VERSION=2.0.0"
-    // returns a <ExceptionText><![CDATA[Missing typeNames parameter]]></ExceptionText>
-    if ( i == 0 && response.indexOf( "<![CDATA[Missing typeNames parameter]]>" ) >= 0 )
-    {
-      QgsDebugMsg( QStringLiteral( "Server does not accept TYPENAME parameter for DescribeFeatureType. Re-trying with TYPENAMES" ) );
-      bUsePlural = true;
-    }
-    // "http://services.cuzk.cz/wfs/inspire-cp-wfs.asp?SERVICE=WFS&REQUEST=DescribeFeatureType&VERSION=2.0.0&TYPENAME=cp:CadastralParcel" returns
-    // <!--Generated by Marushka, version 4.2.5.0, GEOVAP, spol. s r.o., 31.05.2018.-->
-    // <ExceptionReport xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" version="1.0.0" xml:lang="en-US" xmlns="http://www.opengis.net/ows/1.1">
-    // <Exception exceptionCode="OperationProcessingFailed" />
-    // </ExceptionReport>
-    else if ( i == 0 && response.indexOf( "<!--Generated by Marushka" ) >= 0 &&
-              response.indexOf( "OperationProcessingFailed" ) >= 0 )
-    {
-      QgsDebugMsg( QStringLiteral( "Server does not accept TYPENAME parameter for DescribeFeatureType. Re-trying with TYPENAMES" ) );
-      bUsePlural = true;
-    }
-    else
+  QString typeName = mShared->mURI.typeName();
+  QString namespaces = "";
+  Q_FOREACH ( const QgsWfsCapabilities::FeatureType &f, mShared->mCaps.featureTypes )
```

Comment:
I'd say that we should have a method in QgsWfsCapabilites to return the namespaces for a feature type. Would be used here and in qgswfsshareddata.cpp

Reply:
Agree!

Something like this?

`QString namespaces = mShared->mCaps.getNamespaceForTypename(typeName);` where 

__qgswfscapabilities.cpp:__
```
QString QgsWfsCapabilities::Capabilities::getNamespaceForTypename( const QString &name ) const
{
  Q_FOREACH ( const QgsWfsCapabilities::FeatureType &f, featureTypes )
  {
    if ( f.name == name )
    {
      return f.nameSpace;
    }
  }
  return "";
}
```
, and __qgswfscapabilities.h:__
```
struct Capabilities
    {
      Capabilities();

      QString version;
      bool supportsHits;
      bool supportsPaging;
      bool supportsJoins;
      int maxFeatures;
      QList<FeatureType> featureTypes;
      QList<Function> spatialPredicatesList;
      QList<Function> functionList;
      bool useEPSGColumnFormat; // whether to use EPSG:XXXX srsname
      QList< QString > outputFormats;
      QgsStringMap operationGetEndpoints;
      QgsStringMap operationPostEndpoints;

      QSet< QString > setAllTypenames;
      QMap< QString, QString> mapUnprefixedTypenameToPrefixedTypename;
      QSet< QString > setAmbiguousUnprefixedTypename;

      void clear();
      QString addPrefixIfNeeded( const QString &name ) const;
      QString getNamespaceForTypename(const QString &name) const;
    };

```

Reply:
looks good